Dockerfile lint

rpdelaney opened the pull request:
You wrote this, and you're going to tell me you have no idea what docker is about? :P :1st_place_medal:

Commit:
Add apt install --no-install-recommends switch

This prevents apt from installing extra packages that are listed in the
database as "recommended" but aren't dependencies.

Commit:
rm apt-get upgrade

Updates are built in to the base image.

Commit:
Fix broken dpkg-reconfigure argument

Commit:
Add env debianfrontend=noninteractive

This gets rid of some warning messages where debconf progressively falls
back from dialog to readline to noninteractive in the end anyway.

http://manpages.ubuntu.com/manpages/disco/en/man7/debconf.7.html

Commit:
Add env DEBCONF_NONINTERACTIVE_SEEN=true

Reiterating the point that we are not interactive because... debian?
¯\_(ツ)_/¯

Commit:
Coalesce l4d2 user creation into single layer

This just seems logical. No big deal.

Commit:
Dial down curl verbosity

Animated progress meters are hard to read when looking at the build log.

Commit:
Restore --location switch to the right place

ProdigySim commented on Dockerfile:
```diff
@@ -4,38 +4,40 @@ ENV USER l4d2
     && chown $USER:$USER $HOME \
-    && mkdir $SERVER
+    && mkdir $SERVER \
+    && chown --recursive $USER:$USER $SERVER
```
Should we just move the mkdir server down to under the `USER $USER` section?

rpdelaney replied:
That would work. I think `useradd` has options that should obviate the `chown`ing of home directories also. My work day just started but I'll look into that more later tonight.

rpdelaney replied:
8e4eab32f8cc99d3e9f1c9c2c1415cfbcbd47d25

ProdigySim commented on Dockerfile:
```diff
@@ -4,38 +4,40 @@ ENV USER l4d2
+RUN curl --silent --show-error https://sm.alliedmods.net/smdrop/1.10/sourcemod-1.10.0-git6498-linux.tar.gz | tar xvz -C $SERVER/l4d2/left4dead2
 
-RUN curl -L --output $HOME/tmp.zip https://github.com/SirPlease/L4D2-Competitive-Rework/archive/master.zip \
+RUN curl --silent --show-error --output --location $HOME/tmp.zip https://github.com/SirPlease/L4D2-Competitive-Rework/archive/master.zip \
```
This is supposed to be `--output $HOME/tmp.zip`. Can you move the `--location` earlier in the cmd?

rpdelaney replied:
Oops. 50f5e8a

ProdigySim reviewed (changes requested):
Thanks for cleaning this thing up!

Commit:
Fix brocken --output switch

Commit:
Add l4d2 user with --create-home switch

Now we don't have to chown everything under the sun. Instead, switch to
$USER and create stuff from there.

rpdelaney commented:
When I build this, I'm getting errors about invalid SSL certs or params or something. But I don't think (?) it's about the changes I've made here. Let me know what you think.

Commit:
Set locale only after becoming the $USER

Commit:
ADD -> COPY

ADD is the old, more complex way of doing this. COPY preferred when not
doing weird stuff like archive extraction in a single step.

Further reading:
* https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy

ProdigySim commented on Dockerfile:
```diff
@@ -4,43 +4,38 @@ ENV USER l4d2
 RUN apt-get -y update \
-    && apt-get -y upgrade \
-    && apt-get -y install lib32gcc1 curl net-tools lib32stdc++6 locales unzip lib32z1 screen \
+    && apt-get --no-install-recommends -y install lib32gcc1 curl net-tools lib32stdc++6 locales unzip lib32z1 screen \
```
Tested this out tonight. I got SSL errors on the steamcmd usage. Removing `--no-install-recommends` here fixes that issue.

rpdelaney replied:
OK. So I need to figure out _which_ package in the list of recommends is actually required for us, and explicitly install it, I think.

ProdigySim replied:
It's probably the `ca-certificates` package. But, why can't we just leave recommends in place? This is going to run as an internet-connected server; I think we should install stuff that the OS recommends.

rpdelaney replied:
It's not a big deal either way in this case. What motivates my thinking on this is that it's very often a best practice to know exactly what you're putting in the image you're distributing: what is in "recommends" can change over time and the decisions about what to put in there aren't driven by what makes sense to have in docker. In more widely distributed images, or with something where low cold-start time is valuable, image size is also more of a concern than it probably is here. 

This is also part of why `COPY . .` (which we aren't doing here, thankfully) is widely regarded as an anti-pattern/smell: you don't really know what you're putting in the image you're distributing, but you can be sure you're putting in stuff that doesn't need to be there. So, I don't mind figuring out what's missing (and I think your guess is most likely correct). 

Having said all that, if you prefer to just leave the "recommends" in, I understand completely and we can go that route.

ProdigySim replied:
Thanks for walking me through that. I think we should just leave in the recommends for now. A lot of the external sources could change from under us. I'd rather not miss out on security patches or CA cert revocations because we disabled this.

rpdelaney replied:
f47ece151884feab27aa89fe2a2ff75a4d8b98b2

ProdigySim commented:
Were these changes from some automated tool to lint the dockerfile? Can you link me if so? I'd love to set up those kind of checks on CICD

rpdelaney commented:
Some of these proposed changes were from [this](https://www.fromlatest.io/#/) linter (which also has a CLI distribution) but most were just stuff I've picked up over time. I'm not (by any means) a docker guru at work so I'll check in with them to see if they prefer other linters.

We use [pre-commit](https://pre-commit.com/) to run stuff like this in CI, which is great because that way CI runs the same linters that run in your local dev environment. It's a nice pattern.

Commit:
Merge branch 'master' into dockerfile_lint

Commit:
Revert "Add apt install --no-install-recommends switch"

This reverts commit e477259d9e43ba58c7b054d20c701e6d0c1eb2fa.

- https://github.com/ProdigySim/l4d2-rework-docker/pull/1#discussion_r514019153
